File: ios/Image/Image.m
//
//  Image.m
//  react-native-jjkit
//
//  Created by Juan J LF on 4/20/20.
//

#import <Foundation/Foundation.h>
#import "React/RCTViewManager.h"
#import <React/RCTBridgeModule.h>


@interface


RCT_EXTERN_MODULE(Image,RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(source, NSDictionary)
RCT_EXPORT_VIEW_PROPERTY(scaleType, NSString)

RCT_EXPORT_VIEW_PROPERTY(onLoadStart, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onLoadError, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onLoadSuccess, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onLoadEnd, RCTDirectEventBlock)

RCT_EXTERN_METHOD(requestImageByTag:(nonnull NSNumber)tag format:(NSString)format quality:(CGFloat)quality resolve:(RCTPromiseResolveBlock)resolve
rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(clear:(nonnull NSNumber)tag)

@end

File: ios/Image/Image.swift
//
//  Image.swift
//  react-native-jjkit
//
//  Created by Juan J LF on 4/20/20.
//

import Foundation
import UIKit


@objc(Image)
class Image: RCTViewManager {

    override func view() -> UIView! {
    
       return ImageView()
     }
    override static func requiresMainQueueSetup() -> Bool {
      return true
    }
    
    @objc func requestImageByTag(_ tag:NSNumber,format:String,quality:CGFloat, resolve: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock){
      
         self.bridge?.uiManager?.addUIBlock { (_, views) in
            if let v = views?[tag] as? ImageView{
                
                if let re = format == "png" ? v.image?.pngData() : v.image?.jpegData(compressionQuality: quality){
                    resolve(re.base64EncodedString())
                    
                }else{
                    rejecter("Image", "failed to compress with Format \(format)", nil)
                }
               
               
            }else{
                rejecter("Image", "getImageByTag not found view", nil)
            }
        }
      
    }
    
    
}
